docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,164 @@
+# SDG API - Add a file format for defining custom Flows
+
+## Problem Statement
+
+The `instructlab/sdg` library is introducing more extensive data generation pipelines. To enable customization, we should allow users of the library to provide a configuration file which defines a custom pipeline or extends an existing pipeline.
+
+In terms of the API constructs, a Pipeline is created from a sequence of “block configurations” which express how to instantiate and invoke the individual steps (aka blocks) in the pipeline. A Flow construct serves as a template from which a sequence of block configs can be generated.
+
+## Objective
+
+- Library users can specify a custom flow using a well-defined file format.
+- Library users can either use a custom flow standalone, or combine a custom flow with existing flows.
+- The file format and library can evolve substantially without making breaking changes.
+- Incompatible changes can be introduced while retaining support for existing custom flows for a deprecation period.
+
+## Proposal
+
+We will add an API that instantiates a `Pipeline` object from a list of files:
+
+```python
+pipeline = Pipeline.from_flows(ctx, ['mycustomflow.yaml'])
```

Comment:
It comes from the observation that within the library (in `generate_data()`) we are creating a pipeline that chains together `MMLUBenchFlow` and `SynthKnowledgeFlow`

Reply:
I'm [proposing below](https://github.com/instructlab/dev-docs/pull/109#discussion_r1673815390) to merge `MMLUBenchFlow` into `SynthKnowledgeFlow` so that a pipeline is a directory with `freeform_skills.yaml`, `grounded_skills.yaml`, and `knowledge.yaml`

Reply:
responded below

Reply:
> can you point the code to me? i only see the code have a SDG that chains a list of pipelines, but not a pipeline that chains a list of flows.

I'm just saying we have a sequence of blocks in one file and a sequence of blocks in another file and we're chaining them together

Reply:
But the current abstraction does so by chaining Pipeline in SDG but not multiple flows in Pipeline, right?
I'm simply suggesting to not change this behavior while adding YAML file support.

Reply:
The current API allows `SDG([pipeline1, pipeline2])` or `Pipeline(block_configs + block_configs)` and they are identical. I've been using the latter because "SDG" isn't a noun (it doesn't convey what it does) and because as [pep20 says](https://peps.python.org/pep-0020/) "There should be one-- and preferably only one --obvious way to do it."

This all becomes moot if we merge `MMLUBenchFlow` and `SynthKnowledgeFlow` into a `knowledge.yaml`

Reply:
The semantics between SDG or SyntheticDataGeneration and Pipeline is probably worth another discussion (I think you already have a PR on that). But here I'm only trying to say let's decouple that kind of changes from the YAML file support.
btw feel free to resolve this comment.

Reply:
> This all becomes moot if we merge MMLUBenchFlow and SynthKnowledgeFlow into a knowledge.yaml

I'm interpreting this as we don't actually need support for a list of files under the current proposal of a single `knowledge.yaml`, so I can adjust this to be `from_file()` that loads a single yaml, since that's sufficient for the immediate requirements. We could always add an additional `from_files()` later if it comes up.

Let me know if I got that wrong.

Reply:
(or `from_flow()` .... or `from_yaml()` ... i typoed it, but the point being singular is good enough for the moment)

Reply:
@xukai92 please take a look at the latest version and see if you're good with the changes to this section.